### DISCUSSION TITLE
Are there any large/complex multi page apps to show ?

### DISCUSSION THREAD

**Discussion by PasqualePuzio** [reflex-dev/reflex#1962] [Q&A]
Being Python developers since a very long time and having our current backend entirely written in Python, we're very interested in this framework and we're considering to use it in order to build our next web app which will serve millions of users.

You can imagine our web app as a large SaaS website with several dynamic pages (> 50) and a lot of user interaction.

We were wondering if there are any examples of complex/large multi page web applications written with Pynecone/Reflex currently running in production.

Would you mind to share the links ?

Also, it would be great if you could tell us more about your experience, both in terms of development experience (stability, documentation, support, build time, etc) and user experience (performance, responsiveness, bugs, etc.).

Thanks

**Comment by picklelo** [comment]
The main reflex-web repo for our website is pretty large and should give an example of how larger apps can look: https://github.com/reflex-dev/reflex-web

We are working on optimizations to build time, etc. but the current framework should scale well to a large number of web pages.

**Reply by abulvenz** [reply]
We also plan to use reflex for more than one page. 
So it would be nice if you could share components and states between those different packages (in different directories). 
With the need to subclass when adding substates it is quite hard to do so. 
Let's say I have one packages that contains authentication workflows (living on states) and also components such as a login button and also some pages like a login page.
I could now derive the AppState from AuthState. But what if I plan to use more than one of those packages, that should not be coupled.
What I have seen with that regard is the proposal of rx.LocalState (https://github.com/reflex-dev/reflex/issues/1877) which is nice if the component lives on some global scope.

Would it be possible to have something like:
```python
class AppState(rx.State):
    pass

class SomeOtherSubstate(AppState):
    pass

# adding the AuthState
AppState.add_substate(AuthState)

# and then explicitly add another state inheriting from rx.State
AppState.add_substate(ShoppingCartState)

app = App(state=AppState)
```
As you can see we could have a nice and consistent tree which is defined a bit more explicitly than by mere subclassing.
I have already tried to circumvent the "singleton" checks in `App.__init__()` such that when a base state is explicitly passed it will not complain about other classes derived from `rx.State`. But that alone is not enough, because `State.__init_subclass__` does "some magic for the subclass initialization." :) I'm relatively new to python so that function is a bit hard to understand (switches from class to instance variables and alike). I have already tried to implement such a `rx.State.add_substate(...)` classmethod, which involves setting an explicit parent to the substate class and adding the substate to some explicit substate array on `AppState` and afterwards calling `__init_subclass__` manually to do the wiring. But most of my attempts resulted in infinite recursions...

Another idea, again wishful programming could be to have more than one state tree and if needed, reflex includes a base state inbetween.
```python
app = App(state=[AppState,AuthState,ShoppingCartState])
```

I hope that this is the right discussion for bringing up this point.

Maybe I have overlooked something which already allows for this, that would be the best solution :D

Thank You all for Your work on this awesome framework!

**Reply by masenf** [reply]
My plan is to allow an app to have multiple independent state trees. And the main benefit is having separate change domains. Currently every substate is compiled together into one tree, which makes every stateful component in the app depend on any change in any substate. This results in more rerenders and unnecessary coupling than we would like.

**Reply by PasqualePuzio** [reply]
Per component states would also be useful in my opinion, since most of the time simple components (modals, dialogs, etc.) don't really need to access or change the app state.
